dhall/src/Dhall/ETags.hs

```diff
@@ -0,0 +1,136 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | This module contains the implementation of the @dhall tags@ command
+
+module Dhall.ETags
+    ( -- * ETags
+      generate
+    ) where
+
+import Dhall.Util (Input(..))
+import Dhall.Core (Expr(..), Binding(..))
+import Dhall.Src (Src(..))
+import qualified Dhall.Map as DM (keys)
+import Dhall.Parser
+import Data.Text (Text)
+import Control.Monad
+import Control.Exception
+import qualified Data.Text as T
+import qualified Data.Text.IO as TIO
+import qualified Data.Map as M
+import Data.List (isSuffixOf)
+import System.Directory (doesDirectoryExist, getDirectoryContents,
+#if MIN_VERSION_directory(1,3,0)
+                         pathIsSymbolicLink)
+#else
+                         isSymbolicLink)
+#endif
+import System.FilePath ((</>), takeFileName)
+import Text.Megaparsec (sourceLine, sourceColumn, unPos)
+
+data FilePos = FP Int Int deriving (Eq, Ord)
+newtype Tags = Tags (M.Map FilePath (M.Map FilePos Tag))
+instance Semigroup Tags where
+    (Tags ts1) <> (Tags ts2) = Tags (M.unionWith M.union ts1 ts2)
+data Tag = Tag Text Text
+
+{-| Generate ETags for Dhall expressions
+-}
+generate :: Input -> [Text] -> Bool -> IO Text
+generate StandardInput _ _ = showTags . tags "" <$> TIO.getContents
```

Comment:
Is this actually useful? You can't really navigate to a tag when you don't have the filename, right?

Reply:
I explained how I changed it here:
https://github.com/dhall-lang/dhall-haskell/pull/1398#discussion_r331816317